pkg/endpoint/bpf.go

```diff
@@ -1231,67 +1231,86 @@ func (e *Endpoint) applyPolicyMapChanges() (proxyChanges bool, err error) {
 // difference between the realized and desired policy state without
 // dumping the bpf policy map.
 func (e *Endpoint) syncPolicyMap() error {
-	e.policyDebug(logrus.Fields{
-		"policyRealized": e.realizedPolicy == e.desiredPolicy,
-	}, "syncPolicyMap")
-
-	// Nothing to do if the desired policy is already fully realized.
-	if e.realizedPolicy != e.desiredPolicy {
-		errors := 0
-
-		// Add policy map entries before deleting to avoid transient drops
-		err := e.addPolicyMapDelta()
-		if err != nil {
-			errors++
-		}
-
-		// Delete policy keys present in the realized state, but not present in the desired state
-		for keyToDelete := range e.realizedPolicy.PolicyMapState {
-			// If key that is in realized state is not in desired state, just remove it.
-			if _, ok := e.desiredPolicy.PolicyMapState[keyToDelete]; !ok {
-				if !e.deletePolicyKey(keyToDelete, false, nil) {
-					errors++
-				}
-			}
-		}
-
-		if errors > 0 {
-			return fmt.Errorf("syncPolicyMap failed")
-		}
-	}
-
-	// Still may have changes due to identities added and/or
-	// deleted after the desired policy was computed.
+	// Apply pending policy map changes first so that desired map is up-to-date before
+	// we diff the maps below.
 	_, err := e.applyPolicyMapChanges()
-	return err
-}
+	if err != nil {
+		return err
+	}
 
-// addPolicyMapDelta adds new or updates existing bpf policy map state based
-// on the difference between the realized and desired policy state without
-// dumping the bpf policy map.
-func (e *Endpoint) addPolicyMapDelta() error {
 	// Nothing to do if the desired policy is already fully realized.
 	if e.realizedPolicy == e.desiredPolicy {
+		e.policyDebug(nil, "syncPolicyMap(): not syncing as desired == realized")
 		return nil
 	}
 
+	return e.syncDesiredPolicyMapWith(e.realizedPolicy.PolicyMapState, nil)
+}
+
+// syncDesiredPolicyMapWith updates the bpf policy map state based on the
+// difference between the realized and desired policy state without
+// dumping the bpf policy map.
+func (e *Endpoint) syncDesiredPolicyMapWith(realized policy.MapState, diffs []policy.MapChange) error {
```

Comment:
Instead of passing `diffs` as a parameter, why don't we return the slice? I know that you will say because the usage of the returned `diffs` is optional, but this can be achieved with a boolean:
```
func (e *Endpoint) syncDesiredPolicyMapWith(realized policy.MapState, withDiff bool) (count int, diffs []policy.MapChange, err error) {
```
Which allows us to do [0]

Reply:
You're right, updates to a slice header passed as a parameter would not even be visible to the caller, fixed this. Returning `count`, however is not required, as the returned slice has a length even if it is `nil`.